qcengine/programs/entos.py

```diff
@@ -336,56 +398,134 @@ def parse_output(self, outfiles: Dict[str, str], input_model: "AtomicInput") ->
+
+        # Parse wavefunction quantities from entos_results
+        wavefunction = {}
+        if input_model.protocols.wavefunction != "none":
+
+            # First parse basis set information
+            if "ao_basis" in entos_results.keys():
+                atom_map = [item[0] for item in entos_results["structure"]["__Atoms"]["atoms"]]
+
+                # Each item in electron_shells is a dictionary containing info for one basis function
+                electron_shells_by_center = {}
+                for basis_item in entos_results["ao_basis"]["__Basis"]["electron_shells"]:
+                    center_index = basis_item["center_index"]
+                    normalized_primitives = basis_item["normalized_primitives"]
+
+                    electron_shell_info = {
+                        "angular_momentum": [basis_item["angular_momentum"]],
+                        "harmonic_type": basis_item["function_type"].split("_")[-1],
+                        "exponents": basis_item["exponents"],
+                        "coefficients": basis_item["coefficients"],
+                    }
+                    if center_index not in electron_shells_by_center:
+                        electron_shells_by_center[center_index] = [electron_shell_info]
+                    else:
+                        electron_shells_by_center[center_index].append(electron_shell_info)
+
+                # Construct center_data from electron_shells_by_center
+                # Note: Duplicate atoms will over write each other
+                center_data = {}
+                for i in range(len(electron_shells_by_center)):
+                    basis_center_info = {"electron_shells": electron_shells_by_center[i]}
+                    center_data[atom_map[i]] = basis_center_info
+
+                # Construct BasisSet
+                basis_info = {
+                    "name": input_model.model.basis,
+                    # "description": "", # None provided by entos
+                    "center_data": center_data,
+                    "atom_map": atom_map,
+                    "nbf": entos_results["ao_basis"]["__Basis"]["n_functions"],
+                }
+                basis_set = BasisSet(**basis_info)
+                wavefunction["basis"] = basis_set
+            else:
+                raise KeyError(
+                    f"Basis set information not found so wavefunction protocol {input_model.protocols.wavefunction} is not available."
+                )
+
+            # Now parse wavefunction information
+            n_channels = entos_results["n_channels"]
+            if n_channels == 1:
+                wavefunction["restricted"] = True
+                for key in wavefunction_map["restricted"].keys():
+                    if key in entos_results:
+                        wavefunction[wavefunction_map["restricted"][key]] = entos_results[key]
+                        # TODO Uncomment if CCA ordering should be used
```

Comment:
CCA ordering is always used, can you elaborate on this comment?

Reply:
When I first made this PR, I hadn't reordered the AO functions to follow CCA ordering. So I've been in the process of writing the functions to perform this reordering. I'm happy to put the functions I've just written in a general place or do such general functions already exist in QCEngine?

Reply:
I am now using CCA ordering.